twenty-first/src/shared_math/mpolynomial.rs

```diff
@@ -2370,4 +2387,89 @@ mod test_mpolynomials {
         };
         assert_eq!(6, polynomial.degree());
     }
+
+    #[test]
+    fn mpoly_hashing_is_deterministic_and_unaffected_by_zero_coefficients_test() {
+        let variable_count = 10;
+        let mut rnd_mvpoly = gen_mpolynomial(variable_count, 10, 15, 540);
+        for _ in 0..10 {
+            let mut hasher0 = DefaultHasher::new();
+            rnd_mvpoly.hash(&mut hasher0);
+            let hash0 = hasher0.finish();
+            let mut hasher1 = DefaultHasher::new();
+            rnd_mvpoly.hash(&mut hasher1);
+            let hash1 = hasher1.finish();
+
+            assert_eq!(hash0, hash1, "hashing must be deterministic");
+
+            // Hashing is unchanged by 0-coefficient
+            rnd_mvpoly
+                .coefficients
+                .insert(vec![9; variable_count], BFieldElement::zero());
+            let mut hasher2 = DefaultHasher::new();
+            rnd_mvpoly.hash(&mut hasher2);
+            let hash2 = hasher2.finish();
+            assert_eq!(
+                hash0, hash2,
+                "hashing must be invariant under 0-coefficients"
+            );
+
+            // Digest must change if mpoly changes (non-zero coefficient term)
+            rnd_mvpoly
+                .coefficients
+                .insert(vec![2; variable_count], BFieldElement::one());
+            let mut hasher3 = DefaultHasher::new();
+            rnd_mvpoly.hash(&mut hasher3);
+            let hash3 = hasher3.finish();
+            assert_ne!(hash0, hash3, "hashing must change if mpoly changes");
```

Comment:
minor syntactical note
```suggestion
            assert_ne!(hash0, hash3, "hash digest must change if mpoly changes");
```